File: py/src/aoc2025/day9.py
from itertools import chain


def _parse(input: str) -> list[tuple[int, int]]:
    return [
        (int(line[: (i := line.index(","))]), int(line[i + 1 :]))
        for line in input.splitlines()
        if "," in line
    ]


def _part1(input: str) -> int:
    lines = _parse(input)
    return max(
        (
            (abs(x1 - x2) + 1) * (abs(y1 - y2) + 1)
            for i, (x1, y1) in enumerate(lines)
            for x2, y2 in lines[:i]
        ),
        default=0,
    )


def _part2(input: str) -> int:
    lines = _parse(input)
    return max(
        (
            (abs(x1 - x2) + 1) * (abs(y1 - y2) + 1)
            for i, (x1, y1) in enumerate(lines)
            for x2, y2 in lines[:i]
            if all(
                max(x3, x4) <= min(x1, x2)
                or max(x1, x2) <= min(x3, x4)
                or max(y3, y4) <= min(y1, y2)
                or max(y1, y2) <= min(y3, y4)
                for (x3, y3), (x4, y4) in zip(lines, chain(lines[1:], lines[:1]))
            )
        ),
        default=0,
    )


parts = _part1, _part2

File: kt/src/main/kotlin/com/github/ephemient/aoc2025/Day9.kt
package com.github.ephemient.aoc2025

import java.util.SortedMap
import java.util.TreeMap
import kotlin.math.abs

object Day9 {
    private fun parse(input: String) = input.lines().mapNotNull { line ->
        val (left, right) = line.split(',', limit = 2)
            .takeIf { it.size == 2 }
            ?: return@mapNotNull null
        val x = left.toIntOrNull() ?: return@mapNotNull null
        val y = right.toIntOrNull() ?: return@mapNotNull null
        x to y
    }

    fun part1(input: String): Int = parse(input).pairs().maxOfOrNull { (a, b) ->
        (abs(a.first - b.first) + 1) * (abs(a.second - b.second) + 1)
    } ?: 0

    fun part2(input: String): Int {
        val points = parse(input)

        val xs: SortedMap<Int, MutableList<Pair<Int, Int>>> = TreeMap()
        val ys: SortedMap<Int, MutableList<Pair<Int, Int>>> = TreeMap()
        for ((i, point1) in points.withIndex()) {
            val point2 = points.getOrNull(i + 1) ?: points.first()
            if (point1.first == point2.first) {
                xs.getOrPut(point1.first) { mutableListOf() }
                    .add(minOf(point1.second, point2.second) to maxOf(point1.second, point2.second))
            }
            if (point1.second == point2.second) {
                ys.getOrPut(point1.second) { mutableListOf() }
                    .add(minOf(point1.first, point2.first) to maxOf(point1.first, point2.first))
            }
        }

        return points.pairs().fold(0) { acc, (left, right) ->
            val area = (abs(left.first - right.first) + 1) * (abs(left.second - right.second) + 1)
            if (acc < area &&
                xs.doesNotIntersect(left.first, left.second, right.first, right.second) &&
                ys.doesNotIntersect(left.second, left.first, right.second, right.first)
            ) area else acc
        }
    }

    private fun <T> List<T>.pairs() = sequence {
        for ((i, y) in this@pairs.withIndex()) {
            for (x in this@pairs.subList(0, i)) {
                yield(x to y)
            }
        }
    }

    private fun SortedMap<Int, out Iterable<Pair<Int, Int>>>.doesNotIntersect(
        key1: Int, value1: Int,
        key2: Int, value2: Int,
    ): Boolean = key1 == key2 || subMap(minOf(key1, key2) + 1, maxOf(key1, key2)).all {
        it.value.all { (value3, value4) ->
            maxOf(value1, value2) <= minOf(value3, value4) ||
                    maxOf(value3, value4) <= minOf(value1, value2)
        }
    }
}
